### PR TITLE
nginx: add meta section to modules

### LINKED ISSUE #202267: nginxModules have no metadata, no licensing information
## Issue description

Packages in the nginxModules attribute set have no licensing information and no meta.

### Steps to reproduce

````
nix-repl> nginxModules.lua.meta.license
error: attribute 'meta' missing

       at «string»:1:1:

            1| nginxModules.lua.meta.license
             | ^
````

@Izorkin @ajs124 @thoughtpolice @fpletz @globin

### PR DESCRIPTION
**Pull request by SuperSandro2000**
###### Description of changes

Closes #202267

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/http/nginx/modules.nix [hunk 1]
```diff
@@ -62,7 +67,15 @@ let self = {
...
       sha256 = "0yf34s11vgkcl03wbl6gjngm3p9hs8vvm7hkjkwhjh39vkk2a7cy";
     };
+
     inputs = [ openssl ];
+
+    meta = with lib; {
+      description = "Validates Akamai v2 query string tokens";
+      homepage = "https://github.com/kaltura/nginx-akamai-token-validate-module";
+      license = with licenses; [ bsd2 ];
+      maintainers = with maintainers; [ agpl3 ];
```
Comment (ajs124) [inline pkgs/servers/http/nginx/modules.nix:77]: what?

File: pkgs/servers/http/nginx/modules.nix [hunk 2]
```diff
@@ -62,7 +70,15 @@ let self = {
       rev = "34fd0c94d2c43c642f323491c4f4a226cd83b962";
       sha256 = "0yf34s11vgkcl03wbl6gjngm3p9hs8vvm7hkjkwhjh39vkk2a7cy";
     };
+
     inputs = [ openssl ];
+
+    meta = with lib; {
+      description = "Validates Akamai v2 query string tokens";
+      homepage = "https://github.com/kaltura/nginx-akamai-token-validate-module";
+      license = with licenses; [ bsd2 ];
```
Comment (ajs124) [inline pkgs/servers/http/nginx/modules.nix:76]: https://github.com/kaltura/nginx-akamai-token-validate-module/blob/master/LICENSE

**Review by ajs124 — CHANGES_REQUESTED** [review]

File: pkgs/servers/http/nginx/modules.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/servers/http/nginx/modules.nix]: ups, copy paste erorr

File: pkgs/servers/http/nginx/generic.nix [hunk 3]
```diff
@@ -189,7 +189,8 @@ stdenv.mkDerivation {
   meta = if meta != null then meta else {
     description = "A reverse proxy and lightweight webserver";
     homepage    = "http://nginx.org";
-    license     = licenses.bsd2;
+    license     = [ licenses.bsd2 ]
+      ++ lib.foldr (a: b: a ++ b) [ ] (map (m: m.meta.license) modules);
```
Comment (ajs124) [inline pkgs/servers/http/nginx/generic.nix:193]: Couldn't this just be a `concatMap`?

**Review by ajs124 — COMMENTED** [review]
otherwise diff LGTM. didn't verify the licenses tho.

**Review by RaitoBezarius — APPROVED** [review]
LGTM; didn't verify the licenses neither

File: pkgs/servers/http/nginx/modules.nix [hunk 4]
```diff
@@ -126,6 +162,13 @@ let self = {
       rev = "2.5.1";
       sha256 = "0va4jz36mxj76nmq05n3fgnpdad30cslg7c10vnlhdmmic9vqncd";
     };
+
+    meta = with lib; {
+      description = "Adds ability to purge content from FastCGI, proxy, SCGI and uWSGI caches";
+      homepage = "https://github.com/nginx-modules/ngx_cache_purge";
+      license = with licenses; [ bsd3 ];
```
Comment (mweinelt) [inline pkgs/servers/http/nginx/modules.nix:169]:
```suggestion
      license = with licenses; [ bsd2 ];
```

**Review by mweinelt — COMMENTED** [review]
Looks good. One nit.

**Comment by mweinelt** [comment]
nginx passthru tests are failing

**Comment by SuperSandro2000** [comment]
Why are we loading resty core? I don't think the behavior of the test changed in this PR, so that must have been already broken.
```
machine # [   48.210995] nginx[1089]: nginx: [alert] detected a LuaJIT version which is not OpenResty's; many optimizations will be disabled and performance will be compromised (see https://github.com/openresty/luajit2 for OpenResty's LuaJIT or, even better, consider using the OpenResty releases from https://openresty.org/en/download.html)
machine # [   48.214071] nginx[1089]: nginx: [alert] failed to load the 'resty.core' module (https://github.com/openresty/lua-resty-core); ensure you are using an OpenResty release from https://openresty.org/en/download.html (reason: module 'resty.core' not found:
machine # [   48.217150] nginx[1089]:   no field package.preload['resty.core']
machine # [   48.217819] nginx[1089]:   no file './share/lua/5.1/resty/core.lua'
machine # [   48.218621] nginx[1089]:   no file './resty/core.lua'
machine # [   48.219422] nginx[1089]:   no file './resty/core/init.lua'
machine # [   48.220167] nginx[1089]:   no file './lib/lua/5.1/resty/core.so'
machine # [   48.221001] nginx[1089]:   no file './resty/core.so'
machine # [   48.221697] nginx[1089]:   no file './lib/lua/5.1/loadall.so'
machine # [   48.222500] nginx[1089]:   no file './lib/lua/5.1/resty.so'
machine # [   48.223277] nginx[1089]:   no file './resty.so'
machine # [   48.223995] nginx[1089]:   no file './lib/lua/5.1/loadall.so') in /nix/store/jairm365ajddr947vbz8ygd5gk72rmb5-nginx.conf:50
```

**Comment by SuperSandro2000** [comment]
Merging based on previous approvals and that the test wasn't touched in this PR.

**Comment by zowoq** [comment]
The zstd module is missing meta which seems to break it for some reason. https://github.com/NixOS/nixpkgs/pull/229869

**Comment by SuperSandro2000** [comment]
Missed that because it was added later. Thanks for fixing up!

**Comment by hissssst** [comment]
This commit fails building with
```
error: attribute 'meta' missing

       at /nix/store/gdflzjv86wsbyfz59ddqkir0k2npw95q-source/pkgs/servers/http/nginx/generic.nix:191:24:

          190|     license     = [ licenses.bsd2 ]
          191|       ++ concatMap (m: m.meta.license) modules;
             |                        ^
          192|     platforms   = platforms.all;
```

**Comment by SuperSandro2000** [comment]
See above comment, also please add the whole trace.

**Comment by hissssst** [comment]
I've switched to another commit already. Here's the service config

```
{ config, pkgs, lib, ... }:
{
  services.nginx = {
    enable = true;

    # Comperssions
    recommendedBrotliSettings = true;
    recommendedGzipSettings = true;
    recommendedZstdSettings = true;

    recommendedOptimisation = true;
    recommendedProxySettings = true;
    enableReload = true;

    upstreams.onewindow.servers = { "127.0.0.1:4000" = {}; };

    commonHttpConfig = ''
      server {
          listen 0.0.0.0:80 ;
          listen [::0]:80 ;
          server_name localhost ;
          location / {
              auth_basic           "Login please";
              auth_basic_user_file ${ ./htpasswd };
              proxy_pass http://onewindow;
          }
      }
    '';
  };
}
```

**Pull request merged**